In [1]:
import pandas as pd
import numpy as np
from pyspark.sql import SparkSession
import os

In [2]:
spark = SparkSession \
    .builder \
    .appName("Cleaning") \
    .getOrCreate()

In [3]:
# Reading the excel sheets for the H-1B data, the quarterly sheets need to be merged and have duplicates cleaned/removed
# df = spark.read.format("com.elastacloud.spark.excel").load("LCA_Disclosure_Data_FY2021_Q2.xlsx")

In [4]:
H1B_2021_Q1 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2021_Q1.xlsx')

In [5]:
H1B_2021_Q2 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2021_Q2.xlsx')

In [6]:
H1B_2021_Q3 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2021_Q3.xlsx')

In [7]:
H1B_2021_Q4 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2021_Q4.xlsx')

In [8]:
H1B_2020_Q1 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2020_Q1.xlsx')

In [9]:
H1B_2020_Q2 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2020_Q2.xlsx')

In [10]:
H1B_2020_Q3 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2020_Q3.xlsx')

In [11]:
H1B_2020_Q4 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/LCA_Disclosure_Data_FY2020_Q4.xlsx')

In [12]:
H1B_2019 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/H-1B_Disclosure_Data_FY2018_EOY.xlsx')

In [13]:
H1B_2018 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/H-1B_Disclosure_Data_FY2018_EOY.xlsx')

In [14]:
H1B_2017 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/H-1B_Disclosure_Data_FY17.xlsx')

In [15]:
H1B_2016 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/H-1B_Disclosure_Data_FY16.xlsx')

In [16]:
# Reading the excel sheets for the Green Card data, can remove duplicates if they are present but no quarterly reports here
PERM_2021 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/PERM_Disclosure_Data_FY2021.xlsx')

In [17]:
PERM_2020 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/PERM_Disclosure_Data_FY2020.xlsx')

In [18]:
PERM_2019 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/PERM_FY2019.xlsx')

In [19]:
PERM_2018 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/PERM_Disclosure_Data_FY2018_EOY.xlsx')

In [20]:
PERM_2017 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/PERM_Disclosure_Data_FY17.xlsx')

In [21]:
PERM_2016 = pd.read_excel('https://www.dol.gov/sites/dolgov/files/ETA/oflc/pdfs/PERM_Disclosure_Data_FY16.xlsx')

In [22]:
len(list(H1B_2021_Q1.columns))

96

In [23]:
H1B_2021_Q1.head()

CASE_NUMBER CASE_STATUS RECEIVED_DATE DECISION_DATE  \
0  I-200-20268-843329   Certified    2020-09-24    2020-10-01   
1  I-200-20268-843033   Certified    2020-09-24    2020-10-01   
2  I-200-20268-843527   Certified    2020-09-24    2020-10-01   
3  I-200-20268-844241   Certified    2020-09-24    2020-10-01   
4  I-200-20268-844003   Certified    2020-09-24    2020-10-01   

  ORIGINAL_CERT_DATE VISA_CLASS                      JOB_TITLE SOC_CODE  \
0                NaT       H-1B     Senior Technical Architect  15-1132   
1                NaT       H-1B           Programmer Analyst\t  15-1131   
2                NaT       H-1B             Software Developer  15-1132   
3                NaT       H-1B                 JAVA ANALYST 2  15-1121   
4                NaT       H-1B  SQL DATABASE ADMINISTRATOR II  15-1141   

                           SOC_TITLE FULL_TIME_POSITION  ... WILLFUL_VIOLATOR  \
0  Software Developers, Applications                  Y  ...                N   
1               Computer Programmers                  Y  ...                N   
2  Software Developers, Applications                  Y  ...                N   
3          Computer Systems Analysts                  Y  ...                N   
4            Database Administrators                  Y  ...                N   

  SUPPORT_H1B  STATUTORY_BASIS  APPENDIX_A_ATTACHED  PUBLIC_DISCLOSURE  \
0           Y             Wage                  NaN  Disclose Business   
1           Y             Wage                  NaN  Disclose Business   
2           Y             Wage                  NaN  Disclose Business   
3           Y             Wage                  NaN  Disclose Business   
4           Y             Wage                  NaN  Disclose Business   

   PREPARER_LAST_NAME  PREPARER_FIRST_NAME  PREPARER_MIDDLE_INITIAL  \
0            Chothani               Poorvi                        R   
1                 NaN                  NaN                      NaN   
2                 NaN                  NaN                      NaN   
3                 NaN                  NaN                      NaN   
4                 NaN                  NaN                      NaN   

                    PREPARER_BUSINESS_NAME          PREPARER_EMAIL  
0  The Law office of Poorvi Chothani, PLLC  gov@poorvichothani.com  
1                                      NaN                     NaN  
2                                      NaN                     NaN  
3                                      NaN                     NaN  
4                                      NaN                     NaN  

[5 rows x 96 columns]

In [24]:
# Getting important columns 
h1b_quarterly_cols = ['CASE_NUMBER', 'CASE_STATUS', 'RECEIVED_DATE', 'DECISION_DATE',
                      'VISA_CLASS', 'JOB_TITLE', 'SOC_CODE','SOC_TITLE', 'FULL_TIME_POSITION', 
                     'EMPLOYER_NAME','EMPLOYER_CITY', 'EMPLOYER_STATE', 'WAGE_RATE_OF_PAY_FROM',
                      'WAGE_UNIT_OF_PAY', 'PREVAILING_WAGE', 'PW_UNIT_OF_PAY']

In [25]:
# create trimmed quarterly reports
H1B_2021_Q1_trim = H1B_2021_Q1[h1b_quarterly_cols]
H1B_2021_Q2_trim = H1B_2021_Q2[h1b_quarterly_cols]
H1B_2021_Q3_trim = H1B_2021_Q3[h1b_quarterly_cols]
H1B_2021_Q4_trim = H1B_2021_Q4[h1b_quarterly_cols]
H1B_2020_Q1_trim = H1B_2020_Q1[h1b_quarterly_cols]
H1B_2020_Q2_trim = H1B_2020_Q2[h1b_quarterly_cols]
H1B_2020_Q3_trim = H1B_2020_Q3[h1b_quarterly_cols]
H1B_2020_Q4_trim = H1B_2020_Q4[h1b_quarterly_cols]

In [26]:
# rename case submitted col to received date
H1B_2016 = H1B_2016.rename(columns={'CASE_SUBMITTED': 'RECEIVED_DATE'})
H1B_2017 = H1B_2017.rename(columns={'CASE_SUBMITTED': 'RECEIVED_DATE'})
H1B_2018 = H1B_2018.rename(columns={'CASE_SUBMITTED': 'RECEIVED_DATE'})
H1B_2019 = H1B_2019.rename(columns={'CASE_SUBMITTED': 'RECEIVED_DATE'})

In [27]:
# rename soc name col to soc title
H1B_2016 = H1B_2016.rename(columns={'SOC_NAME': 'SOC_TITLE'})
H1B_2017 = H1B_2017.rename(columns={'SOC_NAME': 'SOC_TITLE'})
H1B_2018 = H1B_2018.rename(columns={'SOC_NAME': 'SOC_TITLE'})
H1B_2019 = H1B_2019.rename(columns={'SOC_NAME': 'SOC_TITLE'})

In [28]:
H1B_2016_trim = H1B_2016[h1b_quarterly_cols]
H1B_2017_trim = H1B_2017[h1b_quarterly_cols]
H1B_2018_trim = H1B_2018[h1b_quarterly_cols]
H1B_2019_trim = H1B_2019[h1b_quarterly_cols]

In [29]:
# append the reports that are not quarterly
H1B_combined = H1B_2016_trim.append([H1B_2017_trim, H1B_2018_trim, H1B_2019_trim])

In [30]:
# check all rows moved
len(H1B_combined) == len(H1B_2018_trim) + len(H1B_2017_trim) + len(H1B_2016_trim) + len(H1B_2019_trim)

True

In [31]:
H1B_combined.head()

CASE_NUMBER          CASE_STATUS RECEIVED_DATE DECISION_DATE  \
0  I-200-12240-490687  CERTIFIED-WITHDRAWN    2012-08-27    2015-10-16   
1  I-200-13053-847481  CERTIFIED-WITHDRAWN    2013-02-25    2016-01-13   
2  I-200-13088-054259  CERTIFIED-WITHDRAWN    2013-04-23    2015-12-10   
3  I-200-13144-034110  CERTIFIED-WITHDRAWN    2013-05-24    2016-01-19   
4  I-200-13172-415116            WITHDRAWN    2013-06-26    2016-05-20   

  VISA_CLASS                     JOB_TITLE    SOC_CODE  \
0       H-1B  POSTDOCTORAL RESEARCH FELLOW  10-1021.00   
1       H-1B       CHIEF OPERATING OFFICER     11-1011   
2       H-1B         CHIEF PROCESS OFFICER     11-1011   
3       H-1B   REGIONAL PRESIDEN, AMERICAS     11-1011   
4       H-1B  PRESIDENT MONGOLIA AND INDIA     11-1011   

                       SOC_TITLE FULL_TIME_POSITION  \
0  BIOCHEMISTS AND BIOPHYSICISTS                NaN   
1               CHIEF EXECUTIVES                NaN   
2               CHIEF EXECUTIVES                NaN   
3               CHIEF EXECUTIVES                NaN   
4               CHIEF EXECUTIVES                NaN   

                                       EMPLOYER_NAME EMPLOYER_CITY  \
0                             UNIVERSITY OF MICHIGAN     ANN ARBOR   
1                             GOODMAN NETWORKS, INC.         PLANO   
2                          PORTS AMERICA GROUP, INC.   JERSEY CITY   
3  GATES CORPORATION, A WHOLLY-OWNED SUBSIDIARY O...        DENVER   
4                          PEABODY INVESTMENTS CORP.     ST. LOUIS   

  EMPLOYER_STATE  WAGE_RATE_OF_PAY_FROM WAGE_UNIT_OF_PAY  PREVAILING_WAGE  \
0             MI               41000.00             Year         36067.00   
1             TX              400000.00             Year        242674.00   
2             NJ              264000.00             Year        193066.00   
3             CO              220314.00             Year        220314.00   
4             MO                 171.63             Hour            75.73   

  PW_UNIT_OF_PAY  
0           Year  
1           Year  
2           Year  
3           Year  
4           Hour

In [32]:
# append the quarterly reports for cleaning
H1B_quarterly = H1B_2021_Q1_trim.append([H1B_2021_Q2_trim, H1B_2021_Q3_trim, H1B_2021_Q4_trim,
                                         H1B_2020_Q1_trim, H1B_2020_Q2_trim, H1B_2020_Q3_trim, H1B_2020_Q4_trim])

In [33]:
H1B_quarterly.loc[H1B_quarterly["CASE_STATUS"] == "CERTIFIED-WITHDRAWN", "CASE_STATUS"] = "Certified"

In [34]:
H1B_quarterly.head()

CASE_NUMBER CASE_STATUS RECEIVED_DATE DECISION_DATE VISA_CLASS  \
0  I-200-20268-843329   Certified    2020-09-24    2020-10-01       H-1B   
1  I-200-20268-843033   Certified    2020-09-24    2020-10-01       H-1B   
2  I-200-20268-843527   Certified    2020-09-24    2020-10-01       H-1B   
3  I-200-20268-844241   Certified    2020-09-24    2020-10-01       H-1B   
4  I-200-20268-844003   Certified    2020-09-24    2020-10-01       H-1B   

                       JOB_TITLE SOC_CODE                          SOC_TITLE  \
0     Senior Technical Architect  15-1132  Software Developers, Applications   
1           Programmer Analyst\t  15-1131               Computer Programmers   
2             Software Developer  15-1132  Software Developers, Applications   
3                 JAVA ANALYST 2  15-1121          Computer Systems Analysts   
4  SQL DATABASE ADMINISTRATOR II  15-1141            Database Administrators   

  FULL_TIME_POSITION                   EMPLOYER_NAME   EMPLOYER_CITY  \
0                  Y      Hexaware Technologies Inc.          Iselin   
1                  Y                   WIPRO LIMITED  EAST BRUNSWICK   
2                  Y  Mastech Digital InfoTech, Inc.   Moon Township   
3                  Y             VIRTUSA CORPORATION    Southborough   
4                  Y                  XTGLOBAL, INC.           PLANO   

  EMPLOYER_STATE  WAGE_RATE_OF_PAY_FROM WAGE_UNIT_OF_PAY  PREVAILING_WAGE  \
0             NJ               103770.0             Year          97406.0   
1             NJ                84510.0             Year          84510.0   
2             PA                70678.0             Year          70678.0   
3             MA               103376.0             Year          81453.0   
4             TX                91146.0             Year          91146.0   

  PW_UNIT_OF_PAY  
0           Year  
1           Year  
2           Year  
3           Year  
4           Year

In [35]:
len(H1B_quarterly)

1403639

In [36]:
# remove duplicate cases for quarterly
H1B_quarterly_trim = H1B_quarterly.drop_duplicates(subset='CASE_NUMBER', keep="first")
len(H1B_quarterly_trim)

1097206

In [37]:
# remove duplicate cases for non-quarterly
H1B_combined_trim = H1B_combined.drop_duplicates(subset='CASE_NUMBER', keep="first")
len(H1B_combined_trim)

1887063

In [38]:
# rename certified-withdrawn to just certified for h1b combined non-quarterly
H1B_combined.loc[H1B_combined["CASE_STATUS"] == "Certified-Withdrawn", "CASE_STATUS"] = "Certified"

In [39]:
# Create final H1B dataframe
H1B_combined_final = H1B_combined_trim.append(H1B_quarterly_trim)

In [40]:
# exporting to csv
H1B_combined_final.to_csv('H1B_combined_final.csv', index = False)

In [41]:
# selecting the columns for Green Card data
GC_columns = ['CASE_NUMBER', 'CASE_STATUS', 'CASE_RECEIVED_DATE', 'DECISION_DATE', 'REFILE', 'EMPLOYER_NAME', 'EMPLOYER_STATE', 
             'EMPLOYER_CITY', 'EMPLOYER_NUM_EMPLOYEES', 'FW_OWNERSHIP_INTEREST', 'PW_SOC_TITLE', 'PW_AMOUNT_9089', 'PW_UNIT_OF_PAY_9089',
             'PW_LEVEL_9089', 'WAGE_OFFER_FROM_9089', 'WAGE_OFFER_UNIT_OF_PAY_9089', 'JOB_INFO_JOB_TITLE', 'JOB_INFO_EDUCATION',
              'JOB_INFO_EDUCATION_OTHER', 'JOB_INFO_MAJOR', 'JOB_INFO_EXPERIENCE_NUM_MONTHS', 'COUNTRY_OF_CITIZENSHIP', 'FW_INFO_BIRTH_COUNTRY',
              'CLASS_OF_ADMISSION', 'FOREIGN_WORKER_INFO_EDUCATION', 'FW_INFO_EDUCATION_OTHER', 'FOREIGN_WORKER_INFO_MAJOR', 'PW_JOB_TITLE_9089']

In [42]:
#combine Green Card dataframes
GC_combined = PERM_2016.append([PERM_2017, PERM_2018, PERM_2019, PERM_2020, PERM_2021])

In [43]:
#get columns
GC_combined_trim = GC_combined[GC_columns]

In [44]:
# change certified expired to just certified
GC_combined_trim.loc[GC_combined_trim["CASE_STATUS"] == "Certified-Expired", "CASE_STATUS"] = "Certified"

/opt/conda/lib/python3.7/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [45]:
len(GC_combined_trim)

648460

In [46]:
#remove duplicates
GC_combined_final = GC_combined_trim.drop_duplicates(subset='CASE_NUMBER', keep="first")
len(GC_combined_final)

646954

In [47]:
# exporting to csv
GC_combined_final.to_csv('GC_combined_final.csv', index = False)

================================================================================================
# Variable Analysis
================================================================================================

In [1]:
from pyspark.mllib.stat import Statistics
from pyspark.mllib.linalg import Vectors
from pyspark.mllib.classification import LogisticRegressionWithLBFGS, LogisticRegressionModel
from pyspark.mllib.regression import LabeledPoint
from pyspark.ml.feature import VectorAssembler 
from pyspark.mllib.evaluation import MulticlassMetrics
from pyspark.mllib.feature import StandardScaler

import os

from pyspark.sql import SparkSession
spark = SparkSession.builder \
        .master("local") \
        .appName("Final Project") \
        .getOrCreate()
sc = spark.sparkContext

In [2]:
# param init fpr H1B
infileH1B = 'H1B_combined_final.csv'

infileGC = 'GC_combined_final.csv'

 #### We have two data sets. First one is H1B visa category and various fields related to the visa application. Second file is based on Green card application data. We will analyse these two files seperatly. 

## H1B Data Analysis

In [3]:
# read data into dataframe
dfH1B = spark.read.csv(infileH1B, inferSchema=True, header = True)

In [4]:
dfH1B.show(2)

+------------------+-------------------+-------------+-------------+----------+--------------------+----------+--------------------+------------------+--------------------+-------------+--------------+---------------------+----------------+---------------+--------------+
|       CASE_NUMBER|        CASE_STATUS|RECEIVED_DATE|DECISION_DATE|VISA_CLASS|           JOB_TITLE|  SOC_CODE|           SOC_TITLE|FULL_TIME_POSITION|       EMPLOYER_NAME|EMPLOYER_CITY|EMPLOYER_STATE|WAGE_RATE_OF_PAY_FROM|WAGE_UNIT_OF_PAY|PREVAILING_WAGE|PW_UNIT_OF_PAY|
+------------------+-------------------+-------------+-------------+----------+--------------------+----------+--------------------+------------------+--------------------+-------------+--------------+---------------------+----------------+---------------+--------------+
|I-200-12240-490687|CERTIFIED-WITHDRAWN|   2012-08-27|   2015-10-16|      H-1B|POSTDOCTORAL RESE...|10-1021.00|BIOCHEMISTS AND B...|              null|UNIVERSITY OF MIC...|    ANN ARBO

In [6]:
dfH1B.printSchema()

root
 |-- CASE_NUMBER: string (nullable = true)
 |-- CASE_STATUS: string (nullable = true)
 |-- RECEIVED_DATE: string (nullable = true)
 |-- DECISION_DATE: string (nullable = true)
 |-- VISA_CLASS: string (nullable = true)
 |-- JOB_TITLE: string (nullable = true)
 |-- SOC_CODE: string (nullable = true)
 |-- SOC_TITLE: string (nullable = true)
 |-- FULL_TIME_POSITION: string (nullable = true)
 |-- EMPLOYER_NAME: string (nullable = true)
 |-- EMPLOYER_CITY: string (nullable = true)
 |-- EMPLOYER_STATE: string (nullable = true)
 |-- WAGE_RATE_OF_PAY_FROM: string (nullable = true)
 |-- WAGE_UNIT_OF_PAY: string (nullable = true)
 |-- PREVAILING_WAGE: string (nullable = true)
 |-- PW_UNIT_OF_PAY: string (nullable = true)



In [26]:
dfH1B.count()

2984269

H1B dataset has 16 columns listed as above. The column 'CASE_NUMBER' is the unique identification number for each visa application. The dataset has 2,984,269 number of records. 

In [7]:
dfH1B.groupBy("CASE_STATUS").count().show()

+--------------------+-------+
|         CASE_STATUS|  count|
+--------------------+-------+
|              Denied|   6675|
|           CERTIFIED|1694789|
| CERTIFIED-WITHDRAWN| 105135|
|Certified - Withd...|  33083|
|           Withdrawn|  20937|
|           Certified|1036511|
|           WITHDRAWN|  60812|
|              DENIED|  26327|
+--------------------+-------+



The 'CASE_STATUS' indicates the outcome of the H1B application. We will use this variable as the target variable for the logistic and other machine learning models.

Following are frequancy tables for each categorical variables in the dataset.

In [8]:
dfH1B.groupBy("VISA_CLASS").count().show()

+--------------------+-------+
|          VISA_CLASS|  count|
+--------------------+-------+
|         H-1B1 Chile|   5233|
|     H-1B1 Singapore|   5828|
|      E-3 Australian|  57493|
|                H-1B|2915714|
|Select Visa Class...|      1|
+--------------------+-------+



In [9]:
dfH1B.groupBy("JOB_TITLE").count().show()

+--------------------+-----+
|           JOB_TITLE|count|
+--------------------+-----+
|PRACTICE LEAD-FAC...|    1|
|DIRECTOR OF STRAT...|   10|
|HOSPITALITY OPERA...|    3|
|   PRACTICE DIRECTOR|   40|
|VP, PRODUCT DEVEL...|    2|
|SR. MANAGER, BUSI...|    1|
|TECHNICAL PRODUCT...|   16|
|DIRECTOR, ONLINE ...|    1|
|STRATEGIC BUSINES...|    1|
|ASSOCIATE MANAGER...|    1|
|SENIOR MANAGER, D...|    1|
|PRODUCT MANAGER, ...|    1|
|DIRECTOR, MEDICAL...|    2|
|BUSINESS DEVELOPM...|    6|
|VICE PRESIDENT OF...|    1|
|    IT SALES MANAGER|    7|
|SALES DIRECTOR - ...|    2|
|VICE PRESIDENT, S...|    1|
|DIRECTOR, SOLUTIO...|    8|
|  SENIOR MANAGER, IT|   20|
+--------------------+-----+
only showing top 20 rows



In [10]:
dfH1B.groupBy("SOC_CODE").count().show()

+----------+-----+
|  SOC_CODE|count|
+----------+-----+
|   13-2081|   14|
|   19.1021|    1|
|   43-5111|    2|
|   49-2095|    1|
|     16974|   29|
|     45300|    4|
|15-1054.00|   20|
|33-3021.01|    2|
|53-6041.00|    1|
|29-9012.00|    3|
|17-2041.00| 1953|
|29-9091.00|   87|
|   47-1011|   26|
|     38900|    2|
|17-3022.00|   33|
|47-4099.00|    7|
|29-1069.05|    5|
|13-2011.00|11764|
|     24580|    4|
|29-2041.00|   26|
+----------+-----+
only showing top 20 rows



In [11]:
dfH1B.groupBy("SOC_TITLE").count().show()

+--------------------+-----+
|           SOC_TITLE|count|
+--------------------+-----+
|COMPUTER & INFORM...|    5|
|FOOD SERVICE MANA...|  424|
|PROPERTY, REAL ES...|    2|
|AGENTS AND BUSINE...|  146|
|     COST ESTIMATORS| 2427|
|SOFTWARE QUALITY ...|    4|
|CHEMICAL TECHNICIANS|  179|
|CAREER/TECHNICAL ...|    1|
|INSTRUCTIONAL COO...| 1485|
|MUSIC DIRECTORS A...|  175|
|SYSTEMS ADMINISTR...|    3|
|WOODWORKING MACHI...|    1|
|FOREIGN LANGUATE ...|    1|
|ARCHITECTURAL AND...|    1|
|Computer Hardware...|    5|
|Software Develope...|    7|
|Food Service Mana...|   60|
|      Dermatologists|   20|
|Biofuels Producti...|    1|
|DATABASE ADMINIST...|26312|
+--------------------+-----+
only showing top 20 rows



In [12]:
dfH1B.groupBy("FULL_TIME_POSITION").count().show()

+--------------------+-------+
|  FULL_TIME_POSITION|  count|
+--------------------+-------+
|                null| 647902|
|                LLC"|      1|
|Environmental Eng...|      1|
|                   Y|2290799|
|          27-1014.00|      1|
|Regulatory Affair...|      1|
|                   N|  45563|
|               INC."|      1|
+--------------------+-------+



In [13]:
dfH1B.groupBy("EMPLOYER_CITY").count().show()

+----------------+-----+
|   EMPLOYER_CITY|count|
+----------------+-----+
|OAKBROOK TERRACE|  202|
|       PRINCETON|17722|
|          NOVATO|  400|
|       WATERTOWN|  450|
|       SOUTHLAKE| 1406|
|   N. AMITYVILLE|    1|
|         RARITAN|   36|
|        LOCKHART|    3|
|      CUMBERLAND|   88|
|         MARLTON|   90|
|            ANNA|    2|
|       ELLINGTON|   18|
|         EDMONDS|   28|
|      SCHAUMBUEG|   17|
|       NESCONSET|    2|
|      CLEAR LAKE|    3|
|          BAXLEY|    3|
|        ST. ROSE|    7|
|        MAYVILLE|    2|
|     MCMINNVILLE|   12|
+----------------+-----+
only showing top 20 rows



In [14]:
dfH1B.groupBy("EMPLOYER_STATE").count().show()

+--------------+------+
|EMPLOYER_STATE| count|
+--------------+------+
|            SC|  9041|
|            AZ| 21446|
|            LA|  6236|
|            MN| 24822|
|            NJ|362534|
|            DC|  9780|
|            OR|  8537|
| Central lslip|     1|
|            VA| 89760|
|          null|   178|
|            RI|  6083|
|            KY| 10475|
|            WY|   723|
|            NH|  4660|
|            MI|116306|
|            NV|  5356|
|            WI| 17321|
|            ID|  3543|
|            CA|537861|
|            CT| 27857|
+--------------+------+
only showing top 20 rows



In [16]:
dfH1B.groupBy("WAGE_UNIT_OF_PAY").count().show()

+----------------+-------+
|WAGE_UNIT_OF_PAY|  count|
+----------------+-------+
|            Year|2792985|
|       Bi-Weekly|    656|
|           Month|   2521|
|            Hour| 187078|
|            null|     74|
|        150000.0|      1|
|         70000.0|      1|
|         88275.0|      1|
|            Week|    950|
|        155106.0|      1|
|              MA|      1|
+----------------+-------+



In [18]:
dfH1B.groupBy("PW_UNIT_OF_PAY").count().show()

+--------------+-------+
|PW_UNIT_OF_PAY|  count|
+--------------+-------+
|          Year|2791861|
|     Bi-Weekly|    362|
|         Month|   1628|
|          Hour| 187659|
|          null|   2157|
|          Week|    600|
|       61360.0|      1|
|       82867.0|      1|
+--------------+-------+



The dataset has two quantitative variables, namely 'WAGE_RATE_OF_PAY_FROM' and 'PREVAILING_WAGE'. Following are discriptive statistics of those variables.

In [27]:
dfH1B.select(["WAGE_RATE_OF_PAY_FROM","PREVAILING_WAGE"]).describe().show()

+-------+---------------------+-----------------+
|summary|WAGE_RATE_OF_PAY_FROM|  PREVAILING_WAGE|
+-------+---------------------+-----------------+
|  count|              2984226|          2982261|
|   mean|    91920.11798390409| 80853.9592955767|
| stddev|    907931.9804660545|591793.8676214326|
|    min|                  0.0|              0.0|
|    max|                 Year|             Year|
+-------+---------------------+-----------------+



Removing nun-numeric values from both variables, we get the following discriptive statistics.

In [33]:
dfH1B.filter(dfH1B.WAGE_RATE_OF_PAY_FROM.rlike('[0-9]') & dfH1B.PREVAILING_WAGE.rlike('[0-9]'))\
.select(["WAGE_RATE_OF_PAY_FROM","PREVAILING_WAGE"]).describe().show()

+-------+---------------------+-----------------+
|summary|WAGE_RATE_OF_PAY_FROM|  PREVAILING_WAGE|
+-------+---------------------+-----------------+
|  count|              2982256|          2982256|
|   mean|    91922.22951846999|80853.96511464765|
| stddev|    908230.1286119057| 591793.966755297|
|    min|                  0.0|              0.0|
|    max|             99999.99|          99997.0|
+-------+---------------------+-----------------+



## Greencard Application Data Analysis

In [34]:
# Read the data
dfGC = spark.read.csv(infileGC, inferSchema=True, header = True)

In [35]:
dfGC.show(2)

+-------------+-----------+------------------+-------------+------+--------------------+--------------+-------------+----------------------+---------------------+--------------------+--------------+-------------------+-------------+--------------------+---------------------------+--------------------+------------------+------------------------+--------------------+------------------------------+----------------------+---------------------+------------------+-----------------------------+-----------------------+-------------------------+-----------------+
|  CASE_NUMBER|CASE_STATUS|CASE_RECEIVED_DATE|DECISION_DATE|REFILE|       EMPLOYER_NAME|EMPLOYER_STATE|EMPLOYER_CITY|EMPLOYER_NUM_EMPLOYEES|FW_OWNERSHIP_INTEREST|        PW_SOC_TITLE|PW_AMOUNT_9089|PW_UNIT_OF_PAY_9089|PW_LEVEL_9089|WAGE_OFFER_FROM_9089|WAGE_OFFER_UNIT_OF_PAY_9089|  JOB_INFO_JOB_TITLE|JOB_INFO_EDUCATION|JOB_INFO_EDUCATION_OTHER|      JOB_INFO_MAJOR|JOB_INFO_EXPERIENCE_NUM_MONTHS|COUNTRY_OF_CITIZENSHIP|FW_INFO_BIRTH_COUN

In [36]:
dfGC.printSchema()

root
 |-- CASE_NUMBER: string (nullable = true)
 |-- CASE_STATUS: string (nullable = true)
 |-- CASE_RECEIVED_DATE: string (nullable = true)
 |-- DECISION_DATE: string (nullable = true)
 |-- REFILE: string (nullable = true)
 |-- EMPLOYER_NAME: string (nullable = true)
 |-- EMPLOYER_STATE: string (nullable = true)
 |-- EMPLOYER_CITY: string (nullable = true)
 |-- EMPLOYER_NUM_EMPLOYEES: string (nullable = true)
 |-- FW_OWNERSHIP_INTEREST: string (nullable = true)
 |-- PW_SOC_TITLE: string (nullable = true)
 |-- PW_AMOUNT_9089: string (nullable = true)
 |-- PW_UNIT_OF_PAY_9089: string (nullable = true)
 |-- PW_LEVEL_9089: string (nullable = true)
 |-- WAGE_OFFER_FROM_9089: string (nullable = true)
 |-- WAGE_OFFER_UNIT_OF_PAY_9089: string (nullable = true)
 |-- JOB_INFO_JOB_TITLE: string (nullable = true)
 |-- JOB_INFO_EDUCATION: string (nullable = true)
 |-- JOB_INFO_EDUCATION_OTHER: string (nullable = true)
 |-- JOB_INFO_MAJOR: string (nullable = true)
 |-- JOB_INFO_EXPERIENCE_NUM_MONTH

In [37]:
dfGC.count()

650816

Greencard application dataset has 28 columns listed as above. The column 'CASE_NUMBER' is the unique identification number for each greencard application. The dataset has 650,816 number of records. 

In [38]:
dfGC.groupBy("CASE_STATUS").count().show()

+--------------------+-----+
|         CASE_STATUS|count|
+--------------------+-----+
|   FILM & TELEVISION|    1|
|TRAINING OR EXPER...|    4|
|Food Service Mana...|    1|
|Software Develope...|    1|
|   or Related Field"|    1|
|orInformation Sys...|    1|
|Elctrcl Engg-Tele...|    1|
|     ComputerAided)"|    2|
|             Bus.Adm|    1|
|      or rltd field"|    2|
|AUTOMOTIVE SERVIC...|    1|
|             Telecom|    1|
|         Engineering|    2|
|                 MIS|    2|
|                math|    1|
|Electronics& Tele...|    1|
|     Squash Coaching|    1|
|         Elect.Engg.|    2|
|                84.0|    9|
| SoftwareEngineering|    2|
+--------------------+-----+
only showing top 20 rows



The 'CASE_STATUS' indicates the outcome of the greencard application. We will use this variable as the target variable for the logistic and other machine learning models.

Following are frequancy tables for each categorical variables in the dataset.

In [39]:
dfGC.groupBy("REFILE").count().show()

+--------------------+-----+
|              REFILE|count|
+--------------------+-----+
|                 E-2|   10|
|        SOUTH AFRICA|    1|
|                 re"|    1|
|             Bus.Adm|    1|
|             Math or|    1|
|         Elect.Engg.|    1|
|          orrel fld"|    2|
|Redhat Linux Admi...|    1|
|               JAPAN|    4|
|          BANGLADESH|    2|
|                 E-3|    3|
|                 TPS|    2|
|    Businessor rel."|    1|
|            Eng(any)|   97|
|             Science|    1|
|          Not in USA|   16|
|            TANZANIA|    1|
|              JORDAN|    2|
|      Math or equiv"|    1|
|                  EE|    1|
+--------------------+-----+
only showing top 20 rows



In [40]:
dfGC.groupBy("EMPLOYER_STATE").count().show()

+--------------------+-----+
|      EMPLOYER_STATE|count|
+--------------------+-----+
|                 E-2|    1|
|        SOUTH AFRICA|    1|
|                  AZ| 2671|
|                  SC| 2750|
|POST GRADUATE DIP...|    1|
|          NEW JERSEY| 6729|
|          orrel fld"|   16|
|BACHELOR OF COMME...|    1|
|          BANGLADESH|    5|
|                  LA| 1183|
|                  MN| 2958|
|NORTHERN MARIANA ...|   85|
|           WISCONSIN|  649|
|THE BENEFICIARY H...|    1|
|        PENNSYLVANIA| 3341|
|Mech. Engg.  or r...|    1|
|                  NJ|24471|
|            ILLINOIS| 4276|
|            MARYLAND| 1313|
|DISTRICT OF COLUMBIA|  306|
+--------------------+-----+
only showing top 20 rows



In [41]:
dfGC.groupBy("EMPLOYER_CITY").count().show()

+----------------+-----+
|   EMPLOYER_CITY|count|
+----------------+-----+
|       PRINCETON| 2242|
|      ABBOTSFORD|  896|
|       SOUTHLAKE|  206|
|          NOVATO|  100|
|       WATERTOWN|  123|
|         MARLTON|   16|
|      CUMBERLAND|   27|
|OAKBROOK TERRACE|   45|
|         Hanover|   37|
|  northbrunswick|    1|
|      CLEAR LAKE|    4|
|     SANDS POINT|    3|
|            ANNA|    1|
|           Tyler|   26|
|         RARITAN|   14|
|        ST. ROSE|    8|
|     BRAITHWAITE|    2|
|        Bluffton|   14|
|       Fairbanks|    5|
| COPPER MOUNTAIN|    7|
+----------------+-----+
only showing top 20 rows



In [42]:
dfGC.groupBy("FW_OWNERSHIP_INTEREST").count().show()

+---------------------+-----+
|FW_OWNERSHIP_INTEREST|count|
+---------------------+-----+
| INDUSTRIAL & SYST...|    1|
| EDUCATIONAL MEASU...|    1|
| ELECTRONICS & COM...|    3|
|    Financial Analyst|    1|
|              BIOLOGY|    1|
| Information Secur...|    1|
|          ENGINEERING|    1|
| AUTOMATIZATION AN...|    1|
| Interpreters and ...|    1|
| Programmer Analys...|    1|
| SOFTWARE DEVELOPE...|    1|
| PURCHASING AGENTS...|    2|
| Compensation and ...|    1|
| MECHANICAL ENGINE...|    1|
| Software Develope...|    1|
| COMMERCE (U.S. EQ...|    1|
|     Chief Executives|    1|
| COMPUTER APPLICATION|    1|
| STRATEGIC INFORMA...|    1|
| COMPUTER APPLICAT...|    1|
+---------------------+-----+
only showing top 20 rows



In [43]:
dfGC.groupBy("PW_SOC_TITLE").count().show()

+--------------------+-----+
|        PW_SOC_TITLE|count|
+--------------------+-----+
|Food Service Mana...|  819|
| DriverSales Workers|   59|
|AFTER HOURS SUPER...|    1|
|FOOD SERVICE MANA...|   17|
|Biofuels Producti...|    1|
|     Roofing Foreman|    2|
|     COST ESTIMATORS|   17|
|  Metal WelderFitter|    1|
|   COOK INDIAN STYLE|    1|
|Secondary School ...|    1|
|Digital Marketing...|    1|
|MUSIC DIRECTORS A...|    1|
|      Dermatologists|    2|
|Principal Wireles...|    1|
|      Content Editor|    1|
|INSTRUCTIONAL COO...|    1|
|Industrial Organi...|    1|
|Food Preparation ...|    2|
|Equipment Acquisi...|    2|
|Account Services ...|    1|
+--------------------+-----+
only showing top 20 rows



In [45]:
dfGC.groupBy("PW_UNIT_OF_PAY_9089").count().show()

+--------------------+------+
| PW_UNIT_OF_PAY_9089| count|
+--------------------+------+
|ELECTRONICS & COM...|     3|
|OPERATIONS RESEAR...|     1|
|         ENGINEERING|     1|
|                Year|436173|
|Industrial Produc...|     1|
|           Bi-Weekly|    25|
|ENGINEERING/ELECT...|     1|
|VERIFICATION ENGI...|     2|
|               Month|    78|
|                Hour|  8025|
|                null|206329|
|COMPUTER SCIENCE ...|     1|
|                Week|   116|
| Internists, General|     1|
|COMPUTER ENGINEERING|     3|
|  Financial Analysts|     1|
|                H-1B|     1|
|     Sales Engineers|     1|
|SOFTWARE DEVELOPE...|     1|
|COMPUTER APPLICAT...|     1|
+--------------------+------+
only showing top 20 rows



In [46]:
dfGC.groupBy("PW_LEVEL_9089").count().show()

+--------------------+------+
|       PW_LEVEL_9089| count|
+--------------------+------+
|         ENGINEERING|    25|
|Programmer Analys...|     1|
|  TEXTILE TECHNOLOGY|     1|
|CHEMICAL ENGINEERING|     2|
|VERIFICATION ENGI...|     2|
|            Level II|139948|
|                null|236149|
|COMPUTER SCIENCE ...|     1|
|            Level IV|104373|
|ELECTRICAL AND EL...|     4|
|ENGINEERING TECHN...|     1|
|Computer and Info...|     4|
|  TELECOMMUNICATIONS|     1|
|           Level III| 77333|
|COMPUTER ENGINEERING|    17|
|  Financial Analysts|     2|
|SOFTWARE DEVELOPE...|     1|
|Multimedia Artist...|     1|
|Computer Systems ...|     1|
|APPLIED COMPUTER ...|     5|
+--------------------+------+
only showing top 20 rows



In [48]:
dfGC.groupBy("WAGE_OFFER_UNIT_OF_PAY_9089").count().show()

+---------------------------+------+
|WAGE_OFFER_UNIT_OF_PAY_9089| count|
+---------------------------+------+
|                       Year|428094|
|                  Bi-Weekly|    47|
|                      Month|   170|
|                       Hour| 16371|
|                       null|205888|
|       BUSINESS ADMINIST...|     1|
|                       Week|   244|
|       BUSINESS ADMINIST...|     1|
+---------------------------+------+



In [49]:
dfGC.groupBy("JOB_INFO_JOB_TITLE").count().show()

+--------------------+-----+
|  JOB_INFO_JOB_TITLE|count|
+--------------------+-----+
|Programmer/Analys...|    3|
|SYSTEMS ADMINISTR...|   77|
|Associate-Interme...|    1|
|Implementation Co...|   26|
|Product Specialis...|    1|
|          PIZZA COOK|  372|
|           Physician|  156|
|Oracle Applicatio...|   25|
|Software Developm...|  122|
|PLM Operations An...|    1|
|.Net Software Dev...|    4|
|Director of Globa...|    1|
|Learning and Deve...|    1|
|    TAEKWONDO MASTER|    9|
|  Oracle Developer I|    2|
|    QA Engineer - II|    2|
|Software Quality ...|    4|
|Field Service Eng...|    1|
|Chef, Japanese Sp...|    9|
|Data/Information ...|    1|
+--------------------+-----+
only showing top 20 rows



In [50]:
dfGC.groupBy("JOB_INFO_EDUCATION").count().show()

+------------------+------+
|JOB_INFO_EDUCATION| count|
+------------------+------+
|       High School|  9615|
|              None| 49607|
|              null|205894|
|       SOUTH KOREA|     1|
|             Other| 13304|
|       Associate's|  3355|
|         Doctorate| 13687|
|        Bachelor's|172540|
|         VENEZUELA|     1|
|          Master's|182812|
+------------------+------+



In [51]:
dfGC.groupBy("JOB_INFO_EDUCATION_OTHER").count().show()

+------------------------+-----+
|JOB_INFO_EDUCATION_OTHER|count|
+------------------------+-----+
|    Bachelor's or equ...|   16|
|    Bachelor's equiva...|    1|
|    M.D Degree (Docto...|    1|
|    Bachelor or Equiv...|    2|
|             J.D. degree|    6|
|    M.D. or professio...|    3|
|    Professional Degr...|    1|
|    2 YEARS OF COLLEG...|    1|
|    Medical Degree (M...|    1|
|    MD or equivalent ...|    1|
|    Juris Doctor degr...|    1|
|    MD or Foreign Deg...|   14|
|    Ph.D. (all but di...|    1|
|    CERTIFICATE IN CO...|    1|
|    4 years of course...|    1|
|    Approved certific...|    2|
|    Bachelor's or equ...|    1|
|    3 Years universit...|    1|
|    PhD or equivalent...|    1|
|    Master of Archite...|    2|
+------------------------+-----+
only showing top 20 rows



In [52]:
dfGC.groupBy("JOB_INFO_MAJOR").count().show()

+--------------------+-----+
|      JOB_INFO_MAJOR|count|
+--------------------+-----+
|Business, Econ., ...|    2|
|Comp Sci, Comp In...|    1|
|Biomed Engg, Chem...|    2|
|           any field|  231|
|Comp Sci and/or E...|    1|
|Computer Science ...|    3|
|Computer Science,...|    4|
|Computer Science,...|    7|
|Business Administ...|   89|
|Electronics & Com...|    1|
|Computer Info Sys...|    2|
|Information Manag...|   46|
|Elect./Elec./Comp...|    1|
|Mechanical Engine...|    1|
|Computer Science,...|    7|
|Comp Sci, Comp Ap...|    1|
|Photography Graph...|    1|
|International Shi...|    1|
|Civil Engineering...|    1|
|Construction Mana...|    5|
+--------------------+-----+
only showing top 20 rows



In [54]:
dfGC.groupBy("COUNTRY_OF_CITIZENSHIP").count().show()

+----------------------+-----+
|COUNTRY_OF_CITIZENSHIP|count|
+----------------------+-----+
|               ARMENIA|  309|
|               BAHAMAS|  189|
|          SOUTH AFRICA| 2578|
|              CAMBODIA|  280|
|                 JAPAN| 3925|
|            BANGLADESH| 3508|
|                UGANDA|  146|
|            CAPE VERDE|   12|
|            MAURITANIA|   10|
|              TANZANIA|  116|
|              MALDIVES|    3|
|                JORDAN| 1093|
|               LESOTHO|    3|
|               ALBANIA|  473|
|               BELARUS|  913|
|                 MALTA|   14|
|   ANTIGUA AND BARBUDA|   20|
|             MAURITIUS|   78|
|                KUWAIT|   51|
|             LITHUANIA|  200|
+----------------------+-----+
only showing top 20 rows



In [55]:
dfGC.groupBy("FW_INFO_BIRTH_COUNTRY").count().show()

+---------------------+-----+
|FW_INFO_BIRTH_COUNTRY|count|
+---------------------+-----+
|         SOUTH AFRICA| 1699|
|              ARMENIA|  227|
|              BAHAMAS|  141|
|             CAMBODIA|  191|
|                JAPAN| 2901|
|           BANGLADESH| 2147|
|               UGANDA|  110|
|           CAPE VERDE|    8|
|           MAURITANIA|    8|
|             TANZANIA|  104|
|             MALDIVES|    3|
|               JORDAN|  574|
|              LESOTHO|    8|
|              ALBANIA|  282|
|              BELARUS|  693|
|                MALTA|   10|
|  ANTIGUA AND BARBUDA|   11|
|            MAURITIUS|   65|
|               KUWAIT|  418|
|            LITHUANIA|  132|
+---------------------+-----+
only showing top 20 rows



In [56]:
dfGC.groupBy("CLASS_OF_ADMISSION").count().show()

+------------------+-----+
|CLASS_OF_ADMISSION|count|
+------------------+-----+
|               E-2| 8266|
|              H-1C|    1|
|               E-3| 2284|
|               V-2|    6|
|               TPS| 2136|
|               S-6|    2|
|               G-5|   66|
|               R-2|   26|
|        Not in USA|25862|
|             H-1B1|  677|
|               D-2|    1|
|               H-3|  108|
|               EWI| 2111|
|                 Q|   10|
|               P-1|  410|
|              H-1A|   62|
|               F-2| 2045|
|              null|55107|
|               VWT|  104|
|               C-1|   47|
+------------------+-----+
only showing top 20 rows



In [57]:
dfGC.groupBy("FOREIGN_WORKER_INFO_EDUCATION").count().show()

+-----------------------------+------+
|FOREIGN_WORKER_INFO_EDUCATION| count|
+-----------------------------+------+
|                  High School| 10334|
|                         None| 44533|
|                         null|209225|
|                        Other| 17048|
|                  Associate's|  3617|
|                    Doctorate| 20480|
|                   Bachelor's|160305|
|                     Master's|185274|
+-----------------------------+------+



In [58]:
dfGC.groupBy("FW_INFO_EDUCATION_OTHER").count().show()

+-----------------------+-----+
|FW_INFO_EDUCATION_OTHER|count|
+-----------------------+-----+
|   POST SECONDARY DI...|    3|
|   NATIONAL DIPLOMA ...|    1|
|   DMD DOCTOR OF DEN...|    5|
|   DIPLOMA (EQUIVALE...|    1|
|   3 YEAR BACHELORS;...|    1|
|   FIRST PROFESSIONA...|    1|
|   M.D.  MEDICAL DOCTOR|    1|
|   BACH EQUIV BASED ...|    1|
|   BACHELOR OF DENTA...|    1|
|   DIPLOMA (PLUS EXP...|    1|
|   3 YEARS OF STUDY ...|    1|
|   4 YEARS OF UNIVER...|    1|
|   FOREIGN DEGREE EQ...|   18|
|   3 YEARS OF ACADEM...|    1|
|   BACHELOR'S EQUIVA...|    1|
|   DOCTOR OF VETERIN...|    1|
|   BACH. OF MEDICINE...|    1|
|   COMPLETED POST-SE...|    1|
|   TWO YEARS OF HIGH...|    1|
|   BACHELOR DEGREE (...|    1|
+-----------------------+-----+
only showing top 20 rows



In [59]:
dfGC.groupBy("FOREIGN_WORKER_INFO_MAJOR").count().show()

+-------------------------+-----+
|FOREIGN_WORKER_INFO_MAJOR|count|
+-------------------------+-----+
|     INDUSTRIAL & SYST...|  123|
|     COMPUTER APPLICAT...|    9|
|      CHRISTIAN EDUCATION|   31|
|     INTERNATIONAL BUS...|    1|
|     MASTER OF FINE AR...|    1|
|        PHYSICS EDUCATION|   16|
|          SECONDARY LEVEL|    1|
|     COMMERCE, ACCOUNT...|    1|
|                 ROBOTICS|   93|
|     MANAGEMENT WITH S...|    3|
|     COMPUTER ENGINEER...|    1|
|     ENGINEERING (INST...|   11|
|     ENGINEERING - COM...|    1|
|     MANAGEMENT INFORM...|    1|
|     ELECTICAL ENGINEE...|   11|
|     ENGINEERING MANAG...|    8|
|     ENGINEERING (MANU...|    6|
|     PLASTIC MOLDING P...|    1|
|     ELECTROMAGNETIC C...|    1|
|     BACHELOR OF ENGIN...|    2|
+-------------------------+-----+
only showing top 20 rows



In [60]:
dfGC.groupBy("PW_JOB_TITLE_9089").count().show()

+--------------------+-----+
|   PW_JOB_TITLE_9089|count|
+--------------------+-----+
|Software Develope...|   39|
|Food Service Mana...|  185|
|FOOD SERVICE MANA...|   30|
|Textile Pattern D...|    1|
|           Physician|   16|
|PROGRAMMER ANALYS...|    4|
|      Dispatch Clerk|    1|
|MUSIC DIRECTORS A...|    2|
|   CONCRETE FINISHER|    4|
|     COST ESTIMATORS|   24|
|Music Director an...|    4|
|       Personal Chef|    1|
| SENIOR METALLURGIST|    2|
|Restaurant Manage...|    1|
|PARALEGALS AND LE...|    1|
|Information Syste...|    1|
|Web UI Designer/D...|    1|
|Oracle Applicatio...|    2|
|Chef, Japanese Sp...|    2|
|INSTRUCTIONAL COO...|    3|
+--------------------+-----+
only showing top 20 rows



The dataset three two quantitative variables, namely 'PW_AMOUNT_9089', 'WAGE_OFFER_FROM_9089' and 'JOB_INFO_EXPERIENCE_NUM_MONTHS'. Following are discriptive statistics of those variables.

In [61]:
dfGC.select(["PW_AMOUNT_9089","WAGE_OFFER_FROM_9089","JOB_INFO_EXPERIENCE_NUM_MONTHS"]).describe().show()

+-------+----------------+--------------------+------------------------------+
|summary|  PW_AMOUNT_9089|WAGE_OFFER_FROM_9089|JOB_INFO_EXPERIENCE_NUM_MONTHS|
+-------+----------------+--------------------+------------------------------+
|  count|          444717|              342960|                        254335|
|   mean|82318.6006779178|   88072.46095642494|             32.36903048746923|
| stddev|52088.1830374509|   64435.94536638791|            22.482450770115435|
|    min|   #############|       #############|                           0.0|
|    max|            Year|INFORMATION SYSTE...|                      Master's|
+-------+----------------+--------------------+------------------------------+



Removing nun-numeric values from all three variables, we get the following discriptive statistics.

In [63]:
dfGC.filter(dfGC.PW_AMOUNT_9089.rlike('[0-9]') & dfGC.WAGE_OFFER_FROM_9089.rlike('[0-9]') & dfGC.JOB_INFO_EXPERIENCE_NUM_MONTHS.rlike('[0-9]'))\
.select(["PW_AMOUNT_9089","WAGE_OFFER_FROM_9089","JOB_INFO_EXPERIENCE_NUM_MONTHS"]).describe().show()

+-------+-----------------+--------------------+------------------------------+
|summary|   PW_AMOUNT_9089|WAGE_OFFER_FROM_9089|JOB_INFO_EXPERIENCE_NUM_MONTHS|
+-------+-----------------+--------------------+------------------------------+
|  count|           194643|              194643|                        194643|
|   mean|90065.25384204886|   94824.23267089408|             32.39676227760567|
| stddev|52042.14886913379|   65625.78865821898|            22.429905434953344|
|    min|         1,000.00|        1,000,000.00|                           0.0|
|    max|          99988.0|             99999.9|                          98.0|
+-------+-----------------+--------------------+------------------------------+



Need to remove commas in numbers

In [67]:
from pyspark.sql import functions

dfGC.filter(dfGC.PW_AMOUNT_9089.rlike('[0-9]') & dfGC.WAGE_OFFER_FROM_9089.rlike('[0-9]') & dfGC.JOB_INFO_EXPERIENCE_NUM_MONTHS.rlike('[0-9]'))\
.withColumn("PW_AMOUNT_9089", functions.regexp_replace('PW_AMOUNT_9089',r'[,]',""))\
.withColumn("WAGE_OFFER_FROM_9089", functions.regexp_replace('WAGE_OFFER_FROM_9089',r'[,]',""))\
.withColumn("JOB_INFO_EXPERIENCE_NUM_MONTHS", functions.regexp_replace('JOB_INFO_EXPERIENCE_NUM_MONTHS',r'[,]',""))\
.select(["PW_AMOUNT_9089","WAGE_OFFER_FROM_9089","JOB_INFO_EXPERIENCE_NUM_MONTHS"]).describe().show()

+-------+-----------------+--------------------+------------------------------+
|summary|   PW_AMOUNT_9089|WAGE_OFFER_FROM_9089|JOB_INFO_EXPERIENCE_NUM_MONTHS|
+-------+-----------------+--------------------+------------------------------+
|  count|           194643|              194643|                        194643|
|   mean|91980.21292222217|   99242.88092584857|             32.39676227760567|
| stddev|51142.00776281451|   61108.11882525911|            22.429905434953344|
|    min|             10.0|                10.0|                           0.0|
|    max|         99992.00|            99999.94|                          98.0|
+-------+-----------------+--------------------+------------------------------+

